src/codeflare_sdk/ray/cluster/build_ray_cluster.py

```diff
@@ -0,0 +1,554 @@
+import yaml
+import uuid
+import sys
+import warnings
+import json
+
+FORBIDDEN_CUSTOM_RESOURCE_TYPES = ["GPU", "CPU", "memory"]
+VOLUME_MOUNTS = [
+    V1VolumeMount(
+        mount_path="/etc/pki/tls/certs/odh-trusted-ca-bundle.crt",
+        name="odh-trusted-ca-cert",
+        sub_path="odh-trusted-ca-bundle.crt",
+    ),
+    V1VolumeMount(
+        mount_path="/etc/ssl/certs/odh-trusted-ca-bundle.crt",
+        name="odh-trusted-ca-cert",
+        sub_path="odh-trusted-ca-bundle.crt",
+    ),
+    V1VolumeMount(
+        mount_path="/etc/pki/tls/certs/odh-ca-bundle.crt",
+        name="odh-ca-cert",
+        sub_path="odh-ca-bundle.crt",
+    ),
+    V1VolumeMount(
+        mount_path="/etc/ssl/certs/odh-ca-bundle.crt",
+        name="odh-ca-cert",
+        sub_path="odh-ca-bundle.crt",
+    ),
+]
+
+VOLUMES = [
+    V1Volume(
+        name="odh-trusted-ca-cert",
+        config_map=V1ConfigMapVolumeSource(
+            name="odh-trusted-ca-bundle",
+            items=[V1KeyToPath(key="ca-bundle.crt", path="odh-trusted-ca-bundle.crt")],
+            optional=True,
+        ),
+    ),
+    V1Volume(
+        name="odh-ca-cert",
+        config_map=V1ConfigMapVolumeSource(
+            name="odh-trusted-ca-bundle",
+            items=[V1KeyToPath(key="odh-ca-bundle.crt", path="odh-ca-bundle.crt")],
+            optional=True,
+        ),
+    ),
+]
+
+SUPPORTED_PYTHON_VERSIONS = {
+    "3.9": "quay.io/modh/ray@sha256:0d715f92570a2997381b7cafc0e224cfa25323f18b9545acfd23bc2b71576d06",
+    "3.11": "quay.io/modh/ray@sha256:db667df1bc437a7b0965e8031e905d3ab04b86390d764d120e05ea5a5c18d1b4",
+}
+
+
+# RayCluster/AppWrapper builder function
+def build_ray_cluster(cluster: "codeflare_sdk.ray.cluster.Cluster"):
+    """build_ray_cluster is used for creating a Ray Cluster/AppWrapper dict
+
+    The resource is a dict template which uses Kubernetes Objects for creating metadata, resource requests,
+    specs and containers. The result is sanitised and returned either as a dict or written as a yaml file.
+    """
+    ray_version = "2.35.0"
+
+    # GPU related variables
+    head_gpu_count, worker_gpu_count = head_worker_gpu_count_from_cluster(cluster)
+    head_resources, worker_resources = head_worker_extended_resources_from_cluster(
+        cluster
+    )
+    head_resources = json.dumps(head_resources).replace('"', '\\"')
+    head_resources = f'"{head_resources}"'
+    worker_resources = json.dumps(worker_resources).replace('"', '\\"')
+    worker_resources = f'"{worker_resources}"'
+
+    # Create the Ray Cluster using the V1RayCluster Object
+    resource = {
+        "apiVersion": "ray.io/v1",
+        "kind": "RayCluster",
+        "metadata": get_metadata(cluster),
+        "spec": {
+            "rayVersion": ray_version,
+            "enableInTreeAutoscaling": False,
+            "autoscalerOptions": {
+                "upscalingMode": "Default",
+                "idleTimeoutSeconds": 60,
+                "imagePullPolicy": "IfNotPresent",
```

Comment:
I think it would be preferable if this field was unset/removed. If the user decides to use a `:latest` image tag, `IfNotPresent` will cache the tag and not pull the latest image. Unsetting this field will allow us to cache the image when it's not `:latest`, and pull if otherwise.